molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/EntityTypeCopier.java

```diff
@@ -0,0 +1,245 @@
+    EntityType copy = entityTypeMetadataCopier.copy(original, state);
+    String newId = idGenerator.generateId();
+    copy.setId(newId);
+
+    state.copiedEntityTypes().put(original.getId(), copy);
+    state.originalEntityTypeIds().put(newId, original.getId());
+
+    return copy;
+  }
+
+  private void transformRelations(EntityType entityType, CopyState state) {
+    transformPackage(entityType, state.copiedPackages());
+    transformExtends(entityType, state.copiedEntityTypes());
+    transformRefEntities(entityType, state.copiedEntityTypes());
+    transformMappedBys(entityType, state.copiedAttributes());
+  }
+
+  private EntityType cutDefaultValues(EntityType copy, CopyState state) {
+    stream(copy.getAtomicAttributes())
+        .filter(EntityTypeUtils::isReferenceType)
+        .filter(attr -> attr.getDefaultValue() != null)
+        .forEach(
+            attr -> {
+              state.referenceDefaultValues().put(attr.getIdentifier(), attr.getDefaultValue());
+              attr.setDefaultValue(null);
+            });
+    return copy;
+  }
+
+  private EntityType pasteDefaultValues(EntityType copy, CopyState state) {
+    stream(copy.getAtomicAttributes())
+        .filter(EntityTypeUtils::isReferenceType)
+        .forEach(
+            attr ->
+                attr.setDefaultValue(
+                    state.referenceDefaultValues().getOrDefault(attr.getIdentifier(), null)));
+
+    dataService.getMeta().updateEntityType(copy);
+    return copy;
+  }
+
+  private EntityType persistEntityType(EntityType copy) {
+    dataService.getMeta().addEntityType(copy);
+    return copy;
+  }
+
+  private EntityType copyEntities(EntityType copy, CopyState state) {
+    String originalEntityTypeId = state.originalEntityTypeIds().get(copy.getId());
+    if (!copy.isAbstract()) {
+      dataService
+          .getRepository(originalEntityTypeId)
+          .forEachBatched(batch -> addEntityBatch(copy, state, batch), BATCH_SIZE);
+    }
+    return copy;
+  }
+
+  private void addEntityBatch(EntityType copy, CopyState state, List<Entity> batch) {
+    dataService.add(
+        copy.getId(),
+        batch.stream().map(entity -> new PretendingEntity(entity, state.copiedEntityTypes())));
+  }
+
+  /**
+   * When copying rows from one repository to another, the metadata of these entities will not fit
+   * the copied repository and its references because the metadatas will have different IDs. The
+   * PretendingEntity acts like it's the newly copied entity by returning the metadata of the copied
+   * repository instead of the original.
+   */
+  private class PretendingEntity extends AbstractEntityDecorator {
+
+    private final Map<String, EntityType> copiedEntityTypes;
+
+    PretendingEntity(Entity entity, Map<String, EntityType> copiedEntityTypes) {
+      super(entity);
+      this.copiedEntityTypes = copiedEntityTypes;
+    }
+
+    @Override
+    public EntityType getEntityType() {
+      if (delegate().getEntityType() != null) {
```

Comment:
Check is unnecessary since according to the Entity interface:
```java
  /*
   * @return entity meta data, never null
   */
  EntityType getEntityType();
```